src/qs8-vadd/hvx.c.in

```diff
@@ -0,0 +1,156 @@
+      HVX_Vector vb${int(N/32)}_i16_even = Q6_V_lo_W(vb${int(N/32)}_i16);
+      HVX_Vector vb${int(N/32)}_i16_odd = Q6_V_hi_W(vb${int(N/32)}_i16);
+
+    // vacc = va * va_multiplier + vb * vb_multiplier with widening to 32-bit
+    $for N in range(0, BATCH_TILE, 128):
+      HVX_VectorPair va${int(N/32)}_mul_even = Q6_Vw_vmpyi_VwVh(va_multiplier, va${int(N/32)}_i16_even);
+      HVX_VectorPair va${int(N/32)}_mul_odd = Q6_Vw_vmpyi_VwVh(va_multiplier, va${int(N/32)}_i16_odd);
+      HVX_VectorPair vb${int(N/32)}_mul_even = Q6_Vw_vmpyi_VwVh(vb_multiplier, vb${int(N/32)}_i16_even);
+      HVX_VectorPair vb${int(N/32)}_mul_odd = Q6_Vw_vmpyi_VwVh(vb_multiplier, vb${int(N/32)}_i16_odd);
+
+      HVX_Vector vacc${int(N/32)}_even = Q6_Vw_vadd_VwVw(Q6_V_lo_W(va${int(N/32)}_mul_even), Q6_V_lo_W(vb${int(N/32)}_mul_even));
+      HVX_Vector vacc${int(N/32)}_odd = Q6_Vw_vadd_VwVw(Q6_V_lo_W(va${int(N/32)}_mul_odd), Q6_V_lo_W(vb${int(N/32)}_mul_odd));
+      $if BATCH_TILE > N + 64:
+        HVX_Vector vacc${int((N+32)/32)}_even = Q6_Vw_vadd_VwVw(Q6_V_hi_W(va${int(N/32)}_mul_even), Q6_V_hi_W(vb${int(N/32)}_mul_even));
+        HVX_Vector vacc${int((N+32)/32)}_odd = Q6_Vw_vadd_VwVw(Q6_V_hi_W(va${int(N/32)}_mul_odd), Q6_V_hi_W(vb${int(N/32)}_mul_odd));
+
+    // vacc = vbias + vacc
+    $for N in range(0, BATCH_TILE, 128):
+      vacc${int(N/32)}_even = Q6_Vw_vadd_VwVw(vbias, vacc${int(N/32)}_even);
+      vacc${int(N/32)}_odd = Q6_Vw_vadd_VwVw(vbias, vacc${int(N/32)}_odd);
+      $if BATCH_TILE > N + 64:
+        vacc${int((N+32)/32)}_even = Q6_Vw_vadd_VwVw(vbias, vacc${int((N+32)/32)}_even);
+        vacc${int((N+32)/32)}_odd = Q6_Vw_vadd_VwVw(vbias, vacc${int((N+32)/32)}_odd);
+
+    // vacc = shift and narrow to 16-bit. Add voutput_zero_point
+    $for N in range(0, BATCH_TILE, 128):
+      HVX_Vector vacc${int(N/32)} = Q6_Vh_vasr_VwVwR_sat(vacc${int(N/32)}_odd, vacc${int(N/32)}_even, first_shift);
+      vacc${int(N/32)} = Q6_Vh_vadd_VhVh(voutput_zero_point, Q6_Vh_vasr_VhR(vacc${int(N/32)}, rest_shift));
+      $if BATCH_TILE > N + 64:
+        HVX_Vector vacc${int((N+32)/32)} = Q6_Vh_vasr_VwVwR_sat(vacc${int((N+32)/32)}_odd, vacc${int((N+32)/32)}_even, first_shift);
+        vacc${int((N+32)/32)} = Q6_Vh_vadd_VhVh(voutput_zero_point, Q6_Vh_vasr_VhR(vacc${int((N+32)/32)}, rest_shift));
+
+    // narrow: 16-bit to 8-bit
+    $for N in range(0, BATCH_TILE, 128):
+      $if BATCH_TILE > N + 64:
+        HVX_Vector vout${int(N/32)} = Q6_Vb_vpack_VhVh_sat(vacc${int((N+32)/32)}, vacc${int(N/32)});
+      $else:
+        HVX_Vector vout${int(N/32)} = Q6_Vb_vpack_VhVh_sat(vacc${int(N/32)}, vacc${int(N/32)});
+
+    // minmax
+    $for N in range(0, BATCH_TILE, 128):
+      vout${int(N/32)} = Q6_Vb_vmax_VbVb(voutput_min, vout${int(N/32)});
+      vout${int(N/32)} = Q6_Vb_vmin_VbVb(voutput_max, vout${int(N/32)});
+
+    // store output
+    $for N in range(0, BATCH_TILE, 128):
+      $if BATCH_TILE > N + 96:
```

Comment:
Wouldn't `N + 128 <= BATCH_TILE` be easier to understand? I had to think quite a bit about the current condition.